ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterAggregator.java

```diff
@@ -168,4 +344,29 @@ private StatsSnapshot reduce(StatsSnapshot statsSnapshot) {
     }
     return reducedSnapshot;
   }
+
+  /**
+   * Reduce the given {@link StatsSnapshot} whose first level mapped by PartitionClass to a shallower {@link StatsSnapshot}.
+   * by adding entries belonging to each partition together and aggregating based on partition class. The level of partition
+   * would be removed after reduce completes.
+   * @param statsSnapshot the {@link StatsSnapshot} to be reduced
+   * @return the reduced {@link StatsSnapshot}
+   */
+  static StatsSnapshot reduceByPartitionClass(StatsSnapshot statsSnapshot) {
```

Comment:
Based on the offline discussion, I think we should refactor the StatsSnapshot in a way so that operations like this (consolidating a intermediate level) should be native. That is, the caller shouldn't  have to know the exact structure of the snapshot and conduct the consolidation in a specific way. One possible solution is to add a enum field that indicates the current level to StatsSnapshot. The reduce method can then traverse to the desired level and perform consolidation from there. This way if the structure of the generated StatsSnapshot ever change or we'd like to consolidate a different level in the future we won't have to rewrite our code or introduce duplicate code.

Reply:
Good idea. I might need some time to figure a proper way to do that. Maybe, refactoring `StatsSnapshot` could be done in a separate PR.